src/Components/Server/src/ComponentHub.cs

```diff
@@ -87,26 +102,23 @@ private async Task TerminateCircuitGracefully(CircuitHost circuitHost)
             try
             {
                 Log.CircuitTerminatedGracefully(_logger, circuitHost.CircuitId);
-                _circuitRegistry.PermanentDisconnect(circuitHost);
-                await circuitHost.DisposeAsync();
+                await _circuitRegistry.TerminateAsync(circuitHost.CircuitId);
             }
             catch (Exception e)
             {
                 Log.UnhandledExceptionInCircuit(_logger, circuitHost.CircuitId, e);
             }
-
-            await _circuitRegistry.DisconnectAsync(circuitHost, Context.ConnectionId);
         }
 
-        /// <summary>
-        /// Intended for framework use only. Applications should not call this method directly.
-        /// </summary>
-        public string StartCircuit(string baseUri, string uri)
+        public async ValueTask<string> StartCircuit(string baseUri, string uri)
         {
             if (CircuitHost != null)
             {
+                // This is an error condition and an attempt to bind multiple circuits to a single connection.
+                // We can reject this and terminate the connection.
                 Log.CircuitAlreadyInitialized(_logger, CircuitHost.CircuitId);
```

Comment:
It's not arbitrayr and will happen, because dispose can happen in the background. However, `CircuitId` doesn't throw even if it's disposed, so I think we have nothing to do here.

Reply:
Since you set `CircuitHost` to null in dispose, this is still a nullref possibility

https://github.com/aspnet/AspNetCore/pull/12857/files#diff-8c77562947e4970a8d836d0df8be2b79R162

Reply:
I fixed this in a later commit to use a local variable and prevent the race condition :+1: